[Temporal] Implement TemporalDuration::subtract() according to spec

catamorphism opened the pull request:
#### 0275bcef81ef2b237a08b3c07fd7abadacf57957
<pre>
[Temporal] Implement TemporalDuration::subtract() according to spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=298562">https://bugs.webkit.org/show_bug.cgi?id=298562</a>

Reviewed by Yusuke Suzuki.

Add a new addDurations() operation that TemporalDuration::subtract()
calls, as well as some auxiliary classes and ops:

* PlainYearMonth class (will be elaborated in a future PR)
* balanceISOYearMonth()
* balanceISODate()
* regulateISODate()

Some of the previously-failing tests for Temporal/Duration/prototype/subtract
are now enabled.

* JSTests/stress/temporal-duration.js:
* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/DateConstructor.cpp:
(JSC::makeDate): Deleted.
(JSC::makeTime): Deleted.
(JSC::makeDay): Deleted.
* Source/JavaScriptCore/runtime/DateConstructor.h:
(JSC::makeDay):
(JSC::makeDate):
(JSC::makeTime):
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::isValidISODate):
(JSC::ISO8601::createISODateRecord):
* Source/JavaScriptCore/runtime/ISO8601.h:
* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::balanceISODateOld):
(JSC::epochTimeToEpochYear):
(JSC::epochTimeToMonthInYear):
(JSC::epochTimeToDate):
(JSC::TemporalCalendar::balanceISODate):
(JSC::TemporalCalendar::isoDateAdd):
(JSC::TemporalCalendar::isoDateDifference):
(JSC::TemporalCalendar::balanceISOYearMonth):
(JSC::balanceISODate): Deleted.
* Source/JavaScriptCore/runtime/TemporalCalendar.h:
* Source/JavaScriptCore/runtime/TemporalDuration.cpp:
(JSC::TemporalDuration::toInternalDurationRecordWith24HourDays):
(JSC::TemporalDuration::regulateISODate):
(JSC::TemporalDuration::add const):
(JSC::TemporalDuration::addDurations const):
(JSC::TemporalDuration::subtract const):
* Source/JavaScriptCore/runtime/TemporalDuration.h:
* Source/JavaScriptCore/runtime/TemporalObject.h:

Canonical link: <a href="https://commits.webkit.org/300914@main">https://commits.webkit.org/300914@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c11ed30d7e0bcaee3a6c2a0fb5aa5ff6705fbbf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124204 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43897 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34614 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131034 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76296 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44638 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52497 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94485 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62682 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127158 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35576 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111098 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75073 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34521 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29260 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74519 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116351 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105318 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29481 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133721 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/122735 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51127 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38970 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102962 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51513 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107316 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102764 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/26166 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48121 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26366 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/48031 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19522 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50988 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56763 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/153830 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50435 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39094 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53787 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52110 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
```diff
@@ -291,6 +293,32 @@ static bool balanceISODate(double& year, double& month, double& day)
+    auto epochDays = makeDay(year, month - 1, day);
+    double ms = makeDate(epochDays, 0);
+    int32_t y = epochTimeToEpochYear(ms);
+    int32_t m = epochTimeToMonthInYear(ms) + 1;
```
We should not call `WTF::yearMonthDayFromDays` directly since it is costly.
Please use JSDateMath's cache.

catamorphism replied:
Changed in [bd306a9](https://github.com/WebKit/WebKit/pull/51637/commits/bd306a96d53a53ed9d9c7d68f5b26a28447e5c27)

Constellation reviewed (approved):
r=me

catamorphism committed:
[Temporal] Implement TemporalDuration::subtract() according to spec
https://bugs.webkit.org/show_bug.cgi?id=298562

Reviewed by Yusuke Suzuki.

Add a new addDurations() operation that TemporalDuration::subtract()
calls, as well as some auxiliary classes and ops:

* PlainYearMonth class (will be elaborated in a future PR)
* balanceISOYearMonth()
* balanceISODate()
* regulateISODate()

Some of the previously-failing tests for Temporal/Duration/prototype/subtract
are now enabled.

* JSTests/stress/temporal-duration.js:
* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/DateConstructor.cpp:
(JSC::makeDate): Deleted.
(JSC::makeTime): Deleted.
(JSC::makeDay): Deleted.
* Source/JavaScriptCore/runtime/DateConstructor.h:
(JSC::makeDay):
(JSC::makeDate):
(JSC::makeTime):
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::isValidISODate):
(JSC::ISO8601::createISODateRecord):
* Source/JavaScriptCore/runtime/ISO8601.h:
* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::balanceISODateOld):
(JSC::epochTimeToEpochYear):
(JSC::epochTimeToMonthInYear):
(JSC::epochTimeToDate):
(JSC::TemporalCalendar::balanceISODate):
(JSC::TemporalCalendar::isoDateAdd):
(JSC::TemporalCalendar::isoDateDifference):
(JSC::TemporalCalendar::balanceISOYearMonth):
(JSC::balanceISODate): Deleted.
* Source/JavaScriptCore/runtime/TemporalCalendar.h:
* Source/JavaScriptCore/runtime/TemporalDuration.cpp:
(JSC::TemporalDuration::toInternalDurationRecordWith24HourDays):
(JSC::TemporalDuration::regulateISODate):
(JSC::TemporalDuration::add const):
(JSC::TemporalDuration::addDurations const):
(JSC::TemporalDuration::subtract const):
* Source/JavaScriptCore/runtime/TemporalDuration.h:
* Source/JavaScriptCore/runtime/TemporalObject.h:

Canonical link: https://commits.webkit.org/300914@main
